### PR TITLE
Allow file name templated valid-translation-output files

### PR DESCRIPTION
**Pull request by ugermann**
Allow file name templates for --valid-translation-output. These are essentially format strings that can be interpretted by fmtlib. See config_parser.cpp or --help for valid template parameters. Despite what the git log appears to claim, USE_SENTENCEPIECE is off by default, as in the master branch.

### CODE REVIEW

**Comment by ugermann** [comment]
Pinging @snukky.

File: src/training/validator.cpp [hunk 1]
```diff
@@ -353,6 +355,12 @@ float TranslationValidator::validate(const std::vector<Ptr<ExpressionGraph>>& gr
 
   if(options_->hasAndNotEmpty("valid-translation-output")) {
     fileName = options_->get<std::string>("valid-translation-output");
+    fileName = fmt::format(fileName.c_str(),
+			   fmt::arg("E", state.epochs),
+			   fmt::arg("U", state.batches),
+			   fmt::arg("B", state.batchesEpoch),
+			   fmt::arg("W", state.labelsTotal), // alias for T
+			   fmt::arg("T", state.labelsTotal));
```
Comment (snukky) [inline src/training/validator.cpp]: Please align all arguments horizontally, i.e. +2 spaces in lines 359-363.

File: src/training/validator.h [hunk 2]
```diff
@@ -37,7 +37,8 @@ class ValidatorBase : public TrainingObserver {
 public:
   ValidatorBase(bool lowerIsBetter) : lowerIsBetter_(lowerIsBetter), lastBest_{initScore()} {}
 
-  virtual float validate(const std::vector<Ptr<ExpressionGraph>>& graphs) = 0;
+  virtual float validate(const std::vector<Ptr<ExpressionGraph>>& graphs,
+                         const TrainingState& state) = 0;
```
Comment (snukky) [inline src/training/validator.h]: I think we can use `Ptr<TrainingState>` here, following the general convention?

File: src/training/validator.cpp (hunk 1 above)
Comment (snukky) [inline src/training/validator.cpp]: Do we need both "W" and "T"? One of them is not in the help message.

File: src/common/config_parser.cpp [hunk 3]
```diff
@@ -562,7 +562,13 @@ void ConfigParser::addOptionsValidation(cli::CLIWrapper& cli) {
       "Additional args passed to --valid-script-path. These are inserted"
       " between the script path and the output translation-file path");
   cli.add<std::string>("--valid-translation-output",
-     "Path to store the translation");
+     "(Template for) path to store the translation\n"
+     "    template parameters:\n"
```
Comment (snukky) [inline src/common/config_parser.cpp]: Maybe we can start with the simplest usage followed by a more complex one, for example, "Path to store the translation. Can also be a template, e.g. valid-after-{U}-updates-{T}-tokens.out, with parameters: (...)" or similar.

File: src/common/config_parser.cpp [hunk 4]
```diff
@@ -562,7 +562,13 @@ void ConfigParser::addOptionsValidation(cli::CLIWrapper& cli) {
...
+     "    {E} for epoch\n"
+     "    {B} for # of batches within epoch\n"
+     "    {U} for total # of updates\n"
+     "    {T} for total # of tokens seen\n"
+     "    e.g., valid-after-{U}-updates-{T}-tokens.out");
```
Comment (snukky) [inline src/common/config_parser.cpp]: Unfortunately, the CLI library doesn't seem to support multi-line help messages, so everything will be aligned to the left. I think a one-liner is better in that case

File: src/training/validator.cpp [hunk 5]
```diff
@@ -353,6 +355,12 @@ float TranslationValidator::validate(const std::vector<Ptr<ExpressionGraph>>& gr
 
   if(options_->hasAndNotEmpty("valid-translation-output")) {
     fileName = options_->get<std::string>("valid-translation-output");
+    fileName = fmt::format(fileName.c_str(),
```
Comment (snukky) [inline src/training/validator.cpp]: I know that fmt is included in spdlog, but maybe we can add an explicit `include` at the beginning of the file for clearness?

File: src/training/validator.cpp (hunk 1 above)
Comment (ugermann) [reply src/training/validator.cpp]: Done.

Comment (ugermann) [reply src/training/validator.cpp]: "W" removed.

File: src/common/config_parser.cpp (hunk 3 above)
Comment (ugermann) [reply src/common/config_parser.cpp]: Done.

File: src/common/config_parser.cpp (hunk 4 above)
Comment (ugermann) [reply src/common/config_parser.cpp]: Done.

File: src/training/validator.cpp (hunk 5 above)
Comment (ugermann) [reply src/training/validator.cpp]: I put in a comment in the affected function.

File: src/training/validator.h (hunk 2 above)
Comment (ugermann) [reply src/training/validator.h]: You mean like in ValidatorBase::actAfterLoaded()? Done.

**Comment by ugermann** [comment]
Changes made as requested.

File: src/training/validator.cpp [hunk 6]
```diff
@@ -353,6 +355,13 @@ float TranslationValidator::validate(const std::vector<Ptr<ExpressionGraph>>& gr
 
   if(options_->hasAndNotEmpty("valid-translation-output")) {
     fileName = options_->get<std::string>("valid-translation-output");
+    // The formatting below uses fmtlib, which is included with spdlog
+    // and is included via the logger.
+    fileName = fmt::format(fileName.c_str(),
+                           fmt::arg("E", state->epochs),
+                           fmt::arg("U", state->batches),
+                           fmt::arg("B", state->batchesEpoch),
+                           fmt::arg("T", state->labelsTotal));
```
Comment (snukky) [inline src/training/validator.cpp]: Templating is added to TranslationValidator only, but "valid-translation-output" is also used in BleuValidator. We'd like to have this templating in both.

Comment (ugermann) [reply src/training/validator.cpp]: Done.

**Comment by snukky** [comment]
Thanks. I added a regression test for this new feature.

**Pull request merged**